http/address_watcher.go

```diff
@@ -0,0 +1,119 @@
+	"github.com/rsksmart/liquidity-provider/types"
+	log "github.com/sirupsen/logrus"
+	"math/big"
+)
+
+type BTCAddressWatcher struct {
+	btc           connectors.BTCInterface
+	rsk           connectors.RSKInterface
+	lp            providers.LiquidityProvider
+	calledForUser bool
+	quote         *types.Quote
+}
+
+func NewBTCAddressWatcher(btc connectors.BTCInterface, rsk connectors.RSKInterface, provider providers.LiquidityProvider, q *types.Quote) (*BTCAddressWatcher, error) {
+	watcher := BTCAddressWatcher{
+		btc:           btc,
+		rsk:           rsk,
+		lp:            provider,
+		calledForUser: false,
+		quote:         q,
+	}
+	return &watcher, nil
+}
+
+func (w BTCAddressWatcher) OnNewConfirmation(txHash string, confirmations int64, amount float64) {
+	if !w.calledForUser && confirmations >= int64(w.quote.Confirmations) {
+		_, err := w.performCallForUser()
+		if err != nil {
+			log.Errorf("error calling callForUser. value: %v. error: %v", txHash, err)
+			return
+		}
+		w.calledForUser = true
+	}
+
+	if w.calledForUser && confirmations <= w.rsk.GetRequiredBridgeConfirmations() {
+		_, err := w.performRegisterPegIn(txHash)
+		if err != nil {
+			log.Errorf("error calling registerPegIn. value: %v. error: %v", txHash, err)
+			return
+		}
+	}
+}
+
+func (w BTCAddressWatcher) performCallForUser() (*gethTypes.Transaction, error) {
+	q, err := w.rsk.ParseQuote(w.quote)
+	opt := w.getTxOptions(q.GasLimit.Uint64(), q.Value, q.LiquidityProviderRskAddress)
+	if err != nil {
+		return nil, err
+	}
+	res, err := w.rsk.CallForUser(opt, q)
+	if err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+func (w BTCAddressWatcher) performRegisterPegIn(txHash string) (*gethTypes.Transaction, error) {
+	q, err := w.rsk.ParseQuote(w.quote)
+	opt := w.getTxOptions(q.GasLimit.Uint64(), q.Value, q.LiquidityProviderRskAddress)
+	rawTx, err := w.btc.SerializeTx(txHash)
+	if err != nil {
+		return nil, err
+	}
+	pmt, err := w.btc.SerializePMT(txHash)
+	if err != nil {
+		return nil, err
+	}
+	h, err := w.rsk.HashQuote(w.quote)
+	if err != nil {
+		log.Errorf("error hashing quote: %v", err)
+		return nil, err
+	}
+	hb, err := hex.DecodeString(h)
+	if err != nil {
+		return nil, err
+	}
+	signature, err := w.lp.SignHash(hb)
+	if err != nil {
+		return nil, err
+	}
+	bh, err := w.btc.GetBlockNumberByTx(txHash)
+	if err != nil {
+		log.Errorf("error getting block height: %v", err)
+		return nil, err
+	}
+	tx, err := w.rsk.RegisterPegIn(opt, q, signature, rawTx, pmt, big.NewInt(int64(bh)))
+	if err != nil {
+		return nil, err
+	}
+
+	return tx, nil
+}
+
+func (w BTCAddressWatcher) getTxOptions(gasLimit uint64, value *big.Int, lpRskAddress common.Address) *bind.TransactOpts {
+	tx := gethTypes.NewTx(nil) // TODO: empty transaction is ok at this point?
+	chainId := w.rsk.GetChainId()
+
+	limit := gasLimit + 150000
+
+	signer := func(common.Address, *gethTypes.Transaction) (*gethTypes.Transaction, error) {
+		return w.lp.SignTx(tx, chainId)
+	}
```

Comment:
tx is not needed, the signer function should be lp.SignTx.

Reply:
It doesn't allow you to pass that in, probably because signatures don't match.

Reply:
done